editor/src/document/document_file.rs

```diff
@@ -368,13 +381,23 @@ impl DocumentMessageHandler {
 					.iter()
 					.filter_map(|(key, value)| (!self.graphene_document.layer(key).unwrap().overlay).then(|| (key.clone(), *value)))
 					.collect();
-				self.document_history.push((document.clone_without_overlays(), new_layer_data));
+				self.document_undo_history.push((document.clone_without_overlays(), new_layer_data));
 				Ok(())
 			}
 			None => Err(EditorError::NoTransactionInProgress),
 		}
 	}
 
+	pub fn current_history_hash(&self) -> u64 {
+		// We can use the last state of the document to serve as the hash to compare against
+		// This is useful since when the document is empty the hash will be 0
+		self.document_undo_history.last().map(|(graphene_document, _)| graphene_document.hash()).unwrap_or(0)
```

Comment:
Is this expensive? calling:
```rs
graphene_document.hash()
```
results in just calling:
```rs
self.hasher.finish()
```
does hasher.finish have a time complexity greater than O(1)? This could be replaced with a counter that starts at 0 when saved and increments on a redo or a new change and then decrements on an undo. However, this moves the responsibility of maintaining the counter to several places inside document_file instead of a single source of truth. However, if you think that it will help performance I can change it.

Reply:
Oh, yeah was wondering what sort of hash function you were using… lets say there is a reason for the hashing function containing unsafe code, the "hash" is not "really" as hash, it is more of a hash with some randomnes sprinkled in.
I thought I had removed that impl when transitioning to the uuid based system.
The implementation you are currently using is not a performance concern, but well it kinda relies on a bit of undefined behaviour and `std::mem::transmute_copy`'s.
If there is enough value in the current system, we should keep it, otherwise we should probably replace it with something sensible.
The main issue with the current system is that hash(x) might not be equal to hash(x), meaning x == y does not imply hash(x) == hash(y)

Reply:
You can also take a look at the hash implementation yourself and judge whether you can reconcile this with your conscience. Float are just a pain to deal with

Reply:
I went to implement it as a counter, however, it becomes a problem since once a new change is made to the document after undoing needs to invalidate the counter permanently. There are ways that this can be done but it starts to feel like the incorrect solution. The hash I think is safe to use since I am never using it to compare two different documents. The documents saved in document_undo_history are used immutable so I am using the hash to just identify a single state of a document.

Reply:
It might make sense to change the name from hash to identifier. This would alleviate some of the concerns around the fact that x == y does not imply hash(x) == hash(y)

Reply:
Yeah that's a good idea, I thought I had already done that and then crossed it off my mental todo list. If you want you can append the change to this pr, otherwise I'll create an issue for it.

Reply:
OK. I added those changes plus a note about the identifier to try and prevent misconceptions in the future.